linux: use zstd-based kernel compression for 5.13+

Ma27 opened the pull request:
First attempt[1] was reverted because we used the defaults from the kernel's makefile (using the zstd defaults). Using `-19` gives better results:

| Kernel   | Closure Size with XZ | with ZSTD | increase |
| -------- | -------------------- | --------- | -------- |
| 5.15.158 | 111.0M               | 121.5M    | +8,6%    |
| 6.1.90   | 116.3M               | 127.4M    | +8.7%    |
| 6.6.30   | 129.9M               | 142.4M    | +8.7%    |

[1] 2f04c5f8a3063206c66fa69a346d2776b8074115

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Ma27 committed:
linux: use zstd-based kernel compression for 5.13+

First attempt[1] was reverted because we used the defaults from the
kernel's makefile (using the zstd defaults). Using `-19` gives better
results:

| Kernel   | Closure Size with XZ | with ZSTD | increase |
| -------- | -------------------- | --------- | -------- |
| 5.15.158 | 111.0M               | 121.5M    | +8,6%    |
| 6.1.90   | 116.3M               | 127.4M    | +8.7%    |
| 6.6.30   | 129.9M               | 142.4M    | +8.7%    |

[1] 2f04c5f8a3063206c66fa69a346d2776b8074115

Ma27 commented:
Also cc @alois31 @aviallon .

Curious if you could live with that :) 
Will probably throw this against upstream eventually.

fpletz commented:
The release notes should be changed again, right? https://github.com/NixOS/nixpkgs/commit/2f04c5f8a3063206c66fa69a346d2776b8074115#diff-9c253a17d297f3da51e6ef73a3cb898d879ccf8b0c8613199770917e9387d92cL56

Other than that, LGTM!

Ma27 commented:
@fpletz good point :) 
Will depend a little on if we want to put this into 24.05 though :)

Ma27 commented:
Just reused an existing kernel patch. Rebased it so that it applies on 6.x kernels and added a `Tested-by` and re-submitted it as https://lore.kernel.org/linux-kbuild/20240530210342.1540045-1-maximilian@mbosch.me/T/#u

alyssais commented:
I'd prefer to wait for this to be upstream.  I think it'd be fine to just stick with xz (or less efficient zstd) for older kernels and gradually transition to this with kernel updates.

Ma27 commented:
Fine by me.
Already threw it on my fork to see if something unexpected happened, just to be sure :) 

If accepted by upstream, it'd be fine to apply it to older kernels as done here or are there reasons against it? :)

alyssais commented:
> If accepted by upstream, it'd be fine to apply it to older kernels as done here or are there reasons against it? :)

I guess if it applies there's no harm. :)

alyssais commented:
(You could also probably get it into the stable kernels though — worth asking)
